README.md
<!-- prettier-ignore-start -->

<!-- badges-start -->

[![Black Lives Matter!][badge-blm]][link-blm]
[![Maintenance status][badge-maintenance]][link-repo]
[![Last commit timestamp][badge-last-commit]][link-repo]
[![Open issues][badge-issues]][link-issues]
[![Pull requests][badge-pulls]][link-pulls]
[![codecov][badge-codecov]][link-codecov]
[![Source license][badge-license]][link-license]
[![NPM version][badge-npm]][link-npm]
[![semantic-release][badge-semantic-release]][link-semantic-release]

<!-- badges-end -->

<!-- prettier-ignore-end -->

# babel-plugin-explicit-exports-references

Transforms all internal references to a module's exports such that each reference starts with `module.exports` instead of directly referencing an internal name. This enables easy mocking of specific (exported) functions in Jest with Babel/TypeScript, even when the mocked functions call each other in the same module.

(more description incoming)

## Installation and Usage

```Bash
npm install --save-dev babel-plugin-explicit-exports-references
```

And in your `babel.config.js`:

```typescript
module.exports = {
  plugins: ['explicit-exports-references']
};
```

And finally, run Babel through your toolchain (Webpack, Jest, etc) or manually.
For example:

```Bash
npx babel src --out-dir dist
```

## Motivation

(motivation incoming)

## Documentation

Further documentation can be found under [`docs/`][docs].

## Contributing and Support

**[New issues][choose-new-issue] and [pull requests][pr-compare] are always
welcome and greatly appreciated! 🤩** Just as well, you can [star 🌟 this
project][link-repo] to let me know you found it useful! ✊🏿 Thank you!

See [CONTRIBUTING.md][contributing] and [SUPPORT.md][support] for more
information.

[badge-blm]: https://api.ergodark.com/badges/blm 'Join the movement!'
[link-blm]: https://secure.actblue.com/donate/ms_blm_homepage_2019
[badge-maintenance]:
  https://img.shields.io/maintenance/active/2021
  'Is this package maintained?'
[link-repo]:
  https://github.com/xunnamius/babel-plugin-explicit-exports-references
[badge-last-commit]:
  https://img.shields.io/github/last-commit/xunnamius/babel-plugin-explicit-exports-references
  'When was the last commit to the official repo?'
[badge-issues]:
  https://isitmaintained.com/badge/open/Xunnamius/babel-plugin-explicit-exports-references.svg
  'Number of known issues with this package'
[link-issues]:
  https://github.com/Xunnamius/babel-plugin-explicit-exports-references/issues?q=
[badge-pulls]:
  https://img.shields.io/github/issues-pr/xunnamius/babel-plugin-explicit-exports-references
  'Number of open pull requests'
[link-pulls]:
  https://github.com/xunnamius/babel-plugin-explicit-exports-references/pulls
[badge-codecov]:
  https://codecov.io/gh/Xunnamius/babel-plugin-explicit-exports-references/branch/main/graph/badge.svg?token=HWRIOBAAPW
  'Is this package well-tested?'
[link-codecov]:
  https://codecov.io/gh/Xunnamius/babel-plugin-explicit-exports-references
[package-json]: package.json
[badge-license]:
  https://img.shields.io/npm/l/babel-plugin-explicit-exports-references
  "This package's source license"
[link-license]:
  https://github.com/Xunnamius/babel-plugin-explicit-exports-references/blob/main/LICENSE
[badge-npm]:
  https://api.ergodark.com/badges/npm-pkg-version/babel-plugin-explicit-exports-references
  'Install this package using npm or yarn!'
[link-npm]:
  https://www.npmjs.com/package/babel-plugin-explicit-exports-references
[badge-semantic-release]:
  https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg
  'This repo practices continuous integration and deployment!'
[link-semantic-release]: https://github.com/semantic-release/semantic-release
[side-effects-key]:
  https://webpack.js.org/guides/tree-shaking/#mark-the-file-as-side-effect-free
[exports-main-key]:
  https://github.com/nodejs/node/blob/8d8e06a345043bec787e904edc9a2f5c5e9c275f/doc/api/packages.md#package-entry-points
[tree-shaking]: https://webpack.js.org/guides/tree-shaking
[local-pkg]:
  https://github.com/nodejs/node/blob/8d8e06a345043bec787e904edc9a2f5c5e9c275f/doc/api/packages.md#type
[choose-new-issue]:
  https://github.com/Xunnamius/babel-plugin-explicit-exports-references/issues/new/choose
[pr-compare]:
  https://github.com/Xunnamius/babel-plugin-explicit-exports-references/compare
[contributing]: CONTRIBUTING.md
[support]: .github/SUPPORT.md
[docs]: docs


CHANGELOG.md
This project is switching to a semantic-release based workflow. This document
will be updated afterwards.


test/index.test.ts
import pluginTester from 'babel-plugin-tester';
import plugin from '../src/index';
import { join as makePath } from 'path';

const babelOptions = {
  parserOpts: { strictMode: true },
  plugins: [
    '@babel/plugin-syntax-module-string-names',
    '@babel/plugin-proposal-export-default-from',
    '@babel/plugin-proposal-function-bind',
    '@babel/plugin-transform-typescript'
  ],
  presets: [
    [
      '@babel/preset-env',
      {
        // ? https://github.com/babel/babel-loader/issues/521#issuecomment-441466991
        modules: false,
        // ? https://nodejs.org/en/about/releases
        targets: { node: '10.13.0' }
      }
    ],
    ['@babel/preset-typescript', { allowDeclareFields: true }]
  ]
};

pluginTester({
  pluginName: 'explicit-exports-references',
  plugin,
  babelOptions,
  fixtures: makePath(__dirname, '__fixtures__')
});


src/index.ts
import { name as pkgName } from '../package.json';
import { NodePath, PluginObj, PluginPass } from '@babel/core';
import debugFactory from 'debug';
import * as util from '@babel/types';
import template from '@babel/template';

const debug = debugFactory(`${pkgName}:index`);
let globalScope: NodePath['scope'];

function updateExportRefs(
  path: NodePath<util.Identifier>,
  mode: 'named' | 'default'
): void;
function updateExportRefs(
  path: { from: NodePath<util.Identifier>; to: string },
  mode: 'named' | 'default'
): void;
function updateExportRefs(
  path: NodePath<util.Identifier> | { from: NodePath<util.Identifier>; to: string },
  mode: 'named' | 'default'
): void {
  debug(`mode: ${mode}`);

  // @ts-expect-error: need to discriminate between input types
  const idPath = (path.isIdentifier?.() ? path : path.from) as NodePath<util.Identifier>;
  const localName = idPath.node.name;
  // @ts-expect-error: need to discriminate between input types
  const exportedName = (path.to as string) || localName;
  const globalBinding = globalScope.getBinding(localName);
  const refPaths = [
    ...(globalBinding?.referencePaths || []),
    ...(globalBinding?.constantViolations || [])
  ];

  debug(
    `updating ${refPaths?.length || 0} references to ${mode} export "${localName}"` +
      (exportedName != localName ? ` (exported as "${exportedName}")` : '')
  );

  refPaths?.forEach((refPath) => {
    if (
      !!refPath.find(
        (path) =>
          path.isExportSpecifier() ||
          path.isExportNamespaceSpecifier() ||
          path.isExportDefaultSpecifier()
      )
    ) {
      debug('(an export specifier reference was skipped)');
      return;
    }

    if (!!refPath.find((path) => path.isTSType())) {
      debug('(an TypeScript type reference was skipped)');
      return;
    }

    const wasReplaced = !!(refPath.isIdentifier()
      ? refPath
      : refPath.isAssignmentExpression()
      ? refPath.get('left')
      : undefined
    )?.replaceWith(
      template.expression.ast`module.exports.${mode == 'default' ? mode : exportedName}`
    );

    if (!wasReplaced) debug(`(unsupported reference type "${refPath.type}" was skipped)`);
  });
}

export default function (): PluginObj<PluginPass> {
  return {
    name: 'explicit-exports-references',
    visitor: {
      Program(programPath) {
        globalScope = programPath.scope;
      },
      ExportDefaultDeclaration(exportPath) {
        const declaration = exportPath.get('declaration');
        debug(`encountered default export`);

        if (declaration.isFunctionDeclaration() || declaration.isClassDeclaration()) {
          const id = declaration.get('id') as NodePath<util.Identifier>;
          if (id?.node?.name) updateExportRefs(id, 'default');
          else debug('default declaration is anonymous, ignoring');
        } else debug('(ignored)');
      },
      ExportNamedDeclaration(exportPath) {
        const declaration = exportPath.get('declaration');
        const specifiers = exportPath.get('specifiers');

        if (!declaration.node && !specifiers.length) {
          debug('ignored empty named export declaration');
          return;
        }

        debug(`encountered named export`);

        if (declaration.node) {
          if (declaration.isFunctionDeclaration() || declaration.isClassDeclaration()) {
            updateExportRefs(declaration.get('id') as NodePath<util.Identifier>, 'named');
          } else if (declaration.isVariableDeclaration()) {
            declaration.get('declarations').forEach((declarator) => {
              const id = declarator.get('id');
              if (id.isIdentifier()) updateExportRefs(id, 'named');
              else if (id.isObjectPattern()) {
                id.get('properties').forEach((propPath) => {
                  if (propPath.isObjectProperty()) {
                    const propId = propPath.get('value');
                    if (propId.isIdentifier()) updateExportRefs(propId, 'named');
                  } else if (propPath.isRestElement()) {
                    const arg = propPath.get('argument');
                    if (arg.isIdentifier()) updateExportRefs(arg, 'named');
                  }
                });
              }
            });
          } else debug('(ignored)');
        }

        // ? Later exports take precedence over earlier ones
        specifiers.forEach((specifier) => {
          if (!specifier.isExportSpecifier()) {
            debug(`(ignored export specifier type "${specifier.type}")`);
          } else {
            const local = specifier.get('local');
            const exported = specifier.get('exported');

            debug(`encountered specifier "${local} as ${exported}"`);

            if (exported.isIdentifier()) {
              const exportedName = exported.node.name;
              updateExportRefs(
                {
                  from: local,
                  to: exportedName
                },
                exportedName == 'default' ? 'default' : 'named'
              );
            } else {
              debug(
                '(ignored export specifier because module string names are not supported)'
              );
            }
          }
        });
      }
    }
  };
}


webpack.config.js
// This webpack config is only used for compiling the scripts under
// external-scripts/ and helping transpile src/ => dist/ as dual CJS2+ES2015

const nodeExternals = require('webpack-node-externals');

module.exports = {
    name: 'main',
    mode: 'production',
    target: 'node',
    node: false,

    entry: `${__dirname}/src/index.ts`,

    output: {
        filename: 'index.js',
        path: `${__dirname}/dist`,
        libraryTarget: 'commonjs2',
    },

    externals: [nodeExternals()],

    stats: {
        orphanModules: true,
        providedExports: true,
        usedExports: true,
    },

    resolve: { extensions: ['.ts', '.wasm', '.mjs', '.cjs', '.js', '.json'] },
    module: { rules: [{ test: /\.(ts|js)x?$/, loader: 'babel-loader', exclude: /node_modules/ }] },
    optimization: { usedExports: true },
    ignoreWarnings: [/critical dependency:/i],
};


babel.config.js
// * Every now and then, we adopt best practices from CRA
// * https://tinyurl.com/yakv4ggx

// ? https://nodejs.org/en/about/releases
const NODE_OLDEST_LTS = '10.13.0';

module.exports = {
    parserOpts: { strictMode: true },
    plugins: [
        '@babel/plugin-proposal-export-default-from',
        '@babel/plugin-proposal-function-bind',
        '@babel/plugin-transform-typescript',
    ],
    env: {
        // * Used by Jest and `npm test`
        test: {
            sourceMaps: 'both',
            presets: [
                ['@babel/preset-env', { targets: { node: true }}],
                ['@babel/preset-typescript', { allowDeclareFields: true }],
                // ? We don't care about minification
            ]
        },
        // * Used by `npm run build`
        production: {
            presets: [
                ['@babel/preset-env', {
                    // ? https://github.com/babel/babel-loader/issues/521#issuecomment-441466991
                    //modules: false,
                    // ? https://nodejs.org/en/about/releases
                    targets: { node: NODE_OLDEST_LTS }
                }],
                ['@babel/preset-typescript', { allowDeclareFields: true }],
                // ? Webpack will handle minification
            ]
        },
    }
};
